micromamba/src/update.cpp

```diff
@@ -8,10 +8,108 @@
 
 #include "mamba/api/configuration.hpp"
 #include "mamba/api/update.hpp"
+#include "mamba/api/channel_loader.hpp"
 
+#include "mamba/core/transaction.hpp"
+#include "mamba/core/context.hpp"
+#include "mamba/core/util_os.hpp"
+#include "version.hpp"
+
+extern "C"
+{
+#include "solv/evr.h"
+#include "solv/pool.h"
+#include "solv/conda.h"
+#include "solv/repo.h"
+#include "solv/selection.h"
+#include "solv/solver.h"
+}
 
 using namespace mamba;  // NOLINT(build/namespaces)
 
+void
+update_self(const std::optional<std::string>& version)
+{
+    auto& config = mamba::Configuration::instance();
+    auto& ctx = mamba::Context::instance();
+    config.load();
+
+    mamba::MPool pool;
+    mamba::MultiPackageCache package_caches(ctx.pkgs_dirs);
+
+    auto exp_loaded = load_channels(pool, package_caches, 0);
+    if (!exp_loaded)
+    {
+        throw std::runtime_error(exp_loaded.error().what());
+    }
+
+    pool.create_whatprovides();
+    std::string matchspec;
+    if (!version)
+    {
+        matchspec = fmt::format("micromamba>{}", umamba::version());
+    }
+    else
+    {
+        matchspec = fmt::format("micromamba={}", version.value());
+    }
+
+    auto solvable_ids = pool.select_solvables(pool.matchspec2id(matchspec), true);
+
+    if (solvable_ids.empty())
+    {
+        if (pool.select_solvables(pool.matchspec2id("micromamba")).empty())
+        {
+            throw std::runtime_error(
+                "micromamba not found in the loaded channels. Add 'conda-forge' to your config file.");
+        }
+        else
+        {
+            throw std::runtime_error(
+                "No newer version of micromamba found in the loaded channels.");
+        }
+    }
+
+    std::optional<PackageInfo> latest_micromamba = pool.id2pkginfo(solvable_ids[0]);
+    LOG_WARNING << latest_micromamba.value().url;
+
+    ctx.download_only = true;
+    MTransaction t(pool, { latest_micromamba.value() }, package_caches);
+    auto exp_prefix_data = PrefixData::create(ctx.root_prefix);
+    if (!exp_prefix_data)
+    {
+        // TODO: propagate tl::expected mechanism
+        throw std::runtime_error(exp_prefix_data.error().what());
+    }
+    PrefixData& prefix_data = exp_prefix_data.value();
+    t.execute(prefix_data);
+
+    fs::u8path mamba_exe = get_self_exe_path();
+    fs::u8path mamba_exe_bkup = mamba_exe;
+    mamba_exe_bkup.replace_extension(".bkup");
+
+    fs::u8path cache_path = package_caches.get_extracted_dir_path(latest_micromamba.value())
+                            / latest_micromamba.value().str();
+
+    if (on_win)
+    {
+        fs::rename(mamba_exe, mamba_exe_bkup);
+        fs::copy_file(cache_path / "Library" / "bin" / "micromamba.exe",
+                      mamba_exe,
+                      fs::copy_options::overwrite_existing);
+    }
+    else
+    {
+        fs::copy_file(
```

Comment:
Why no backup on Unix?

Shall we attempt to codesign before overwriting the old micromamba binary?

Reply:
We also have the `bkup` file on Unix, but we don't need the `rename` "dance" because we can overwrite files that are currently used :)

Reply:
Can you help me find the line that does the backup on Unix? oO

Reply:
It was in my head, I swear! :)